### PR TITLE
r/aws_dx_gateway_association: Allowed prefix support

### PR DESCRIPTION
**Pull request by ewbankkit**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

Fixes https://github.com/terraform-providers/terraform-provider-aws/issues/8199.

### CODE REVIEW

**Comment by ewbankkit** [comment]
Current acceptance tests:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAwsDxGatewayAssociation_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -parallel 20 -run=TestAccAwsDxGatewayAssociation_basic -timeout 120m
=== RUN   TestAccAwsDxGatewayAssociation_basic
=== PAUSE TestAccAwsDxGatewayAssociation_basic
=== CONT  TestAccAwsDxGatewayAssociation_basic
--- FAIL: TestAccAwsDxGatewayAssociation_basic (1239.62s)
    testing.go:538: Step 0 error: After applying this step and refreshing, the plan was not empty:
        
        DIFF:
        
        UPDATE: aws_dx_gateway_association.test
          allowed_prefixes.#:          "1" => "0"
          allowed_prefixes.1216997074: "10.255.255.0/28" => ""
        
        STATE:
        
        aws_dx_gateway.test:
          ID = 1a8ca6fb-bbfc-4e3d-ac7f-da8bec989173
          provider = provider.aws
          amazon_side_asn = 65362
          name = terraform-testacc-dxgwassoc-1463777356999477301
        aws_dx_gateway_association.test:
          ID = ga-1a8ca6fb-bbfc-4e3d-ac7f-da8bec989173vgw-09d41462383e45e80
          provider = provider.aws
          allowed_prefixes.# = 1
          allowed_prefixes.1216997074 = 10.255.255.0/28
          dx_gateway_association_id = 5297e438-a084-435a-afd7-b6b4c4dae015
          dx_gateway_id = 1a8ca6fb-bbfc-4e3d-ac7f-da8bec989173
          vpn_gateway_id = vgw-09d41462383e45e80
        
          Dependencies:
            aws_dx_gateway.test
            aws_vpn_gateway_attachment.test
        aws_vpc.test:
          ID = vpc-0c35d0d9035c6ffe2
          provider = provider.aws
          arn = arn:aws:ec2:us-west-2:346386234494:vpc/vpc-0c35d0d9035c6ffe2
          assign_generated_ipv6_cidr_block = false
          cidr_block = 10.255.255.0/28
          default_network_acl_id = acl-0d932fbe72553c2d5
          default_route_table_id = rtb-0f5e45ab5fa9bae1e
          default_security_group_id = sg-05e844d9132964d03
          dhcp_options_id = dopt-249eba40
          enable_classiclink = false
          enable_classiclink_dns_support = false
          enable_dns_hostnames = false
          enable_dns_support = true
          instance_tenancy = default
          ipv6_association_id = 
          ipv6_cidr_block = 
          main_route_table_id = rtb-0f5e45ab5fa9bae1e
          owner_id = 346386234494
          tags.% = 1
          tags.Name = terraform-testacc-dxgwassoc-1463777356999477301
        aws_vpn_gateway.test:
          ID = vgw-09d41462383e45e80
          provider = provider.aws
          amazon_side_asn = 64512
          tags.% = 1
          tags.Name = terraform-testacc-dxgwassoc-1463777356999477301
          vpc_id = vpc-0c35d0d9035c6ffe2
        aws_vpn_gateway_attachment.test:
          ID = vpn-attachment-8937c7e5
          provider = provider.aws
          vpc_id = vpc-0c35d0d9035c6ffe2
          vpn_gateway_id = vgw-09d41462383e45e80
        
          Dependencies:
            aws_vpc.test
            aws_vpn_gateway.test
FAIL
FAIL	github.com/terraform-providers/terraform-provider-aws/aws	1239.658s
GNUmakefile:20: recipe for target 'testacc' failed
make: *** [testacc] Error 1
```

File: aws/resource_aws_dx_gateway_association.go [hunk 1]
```diff
@@ -12,30 +13,47 @@ import (
...
 		Create: resourceAwsDxGatewayAssociationCreate,
 		Read:   resourceAwsDxGatewayAssociationRead,
+		Update: resourceAwsDxGatewayAssociationUpdate,
 		Delete: resourceAwsDxGatewayAssociationDelete,
+		Importer: &schema.ResourceImporter{
+			State: resourceAwsDxGatewayAssociationImport,
+		},
 
 		Schema: map[string]*schema.Schema{
+			"allowed_prefixes": {
```
Comment (bflad) [inline aws/resource_aws_dx_gateway_association.go]: I think it is okay to mark this as `Computed: true` for now, given that it automatically includes the VPC CIDR if no prefixes are provided. 👍

Comment (ewbankkit) [reply aws/resource_aws_dx_gateway_association.go]: Done, I think a missing call to `resourceAwsDxGatewayAssociationRead()` at the end of `resourceAwsDxGatewayAssociationCreate()` doesn't help either 😄.

Comment (ewbankkit) [reply aws/resource_aws_dx_gateway_association.go]:
```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAwsDxGatewayAssociation_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -parallel 20 -run=TestAccAwsDxGatewayAssociation_basic -timeout 120m
=== RUN   TestAccAwsDxGatewayAssociation_basic
=== PAUSE TestAccAwsDxGatewayAssociation_basic
=== CONT  TestAccAwsDxGatewayAssociation_basic
--- PASS: TestAccAwsDxGatewayAssociation_basic (943.05s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	943.077s
```

**Comment by ewbankkit** [comment]
Removed _WIP_.
Ready for review.
Acceptance tests:

```console
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAwsDxGatewayAssociation_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -parallel 20 -run=TestAccAwsDxGatewayAssociation_basic -timeout 120m
=== RUN   TestAccAwsDxGatewayAssociation_basic
=== PAUSE TestAccAwsDxGatewayAssociation_basic
=== CONT  TestAccAwsDxGatewayAssociation_basic
--- PASS: TestAccAwsDxGatewayAssociation_basic (943.05s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	943.077s
$ make testacc TEST=./aws/ TESTARGS='-run=TestAccAwsDxGatewayAssociation_allowedPrefixes'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws/ -v -parallel 20 -run=TestAccAwsDxGatewayAssociation_allowedPrefixes -timeout 120m
=== RUN   TestAccAwsDxGatewayAssociation_allowedPrefixes
=== PAUSE TestAccAwsDxGatewayAssociation_allowedPrefixes
=== CONT  TestAccAwsDxGatewayAssociation_allowedPrefixes
--- PASS: TestAccAwsDxGatewayAssociation_allowedPrefixes (1379.20s)
PASS
ok  	github.com/terraform-providers/terraform-provider-aws/aws	1379.232s
```

File: website/docs/r/dx_gateway_association.html.markdown [hunk 2]
```diff
@@ -32,17 +34,62 @@ resource "aws_dx_gateway_association" "example" {
...
+  name            = "example"
+  amazon_side_asn = "64512"
+}
+
+resource "aws_vpc" "example" {
+  cidr_block = "10.255.255.0/28"
+}
+
+resource "aws_vpn_gateway" "example" {
+  vpc_id = "${aws_vpc.test.id}"
```
Comment (bflad) [inline website/docs/r/dx_gateway_association.html.markdown]:
```suggestion
  vpc_id = "${aws_vpc.example.id}"
```

File: website/docs/r/dx_gateway_association.html.markdown [hunk 3]
```diff
@@ -32,17 +34,62 @@ resource "aws_dx_gateway_association" "example" {
...
+```
+
 ## Argument Reference
 
 The following arguments are supported:
 
-* `dx_gateway_id` - (Required) The ID of the Direct Connect Gateway.
+* `dx_gateway_id` - (Required) The ID of the Direct Connect gateway.
 * `vpn_gateway_id` - (Required) The ID of the VGW with which to associate the gateway.
+* `allowed_prefixes` - (Optional) The Amazon VPC prefixes to advertise to the Direct Connect gateway.
```
Comment (bflad) [inline website/docs/r/dx_gateway_association.html.markdown]: Minor improvements here for clarification:

```suggestion
* `allowed_prefixes` - (Optional) VPC prefixes (CIDRs) to advertise to the Direct Connect gateway. Defaults to the CIDR block of the VPC associated with the Virtual Gateway. To enable drift detection, must be configured.
```

**Comment by ewbankkit** [comment]
@bflad Yes, when I ran the tests in parallel in my account I regularly got timeouts on the gateway disassociation but no problems when running serially.
Thanks for cleaning up the doc. Copy pasta.

**Comment by nywilken** [comment]
This has been released in [version 2.8.0 of the Terraform AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#280-april-26-2019). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

**Pull request merged**